lib/Semantics/check-case.cpp

```diff
@@ -0,0 +1,244 @@
+//===-- lib/Semantics/check-case.cpp --------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "check-case.h"
+#include "flang/Common/idioms.h"
+#include "flang/Common/reference.h"
+#include "flang/Evaluate/fold.h"
+#include "flang/Evaluate/type.h"
+#include "flang/Parser/parse-tree.h"
+#include "flang/Semantics/semantics.h"
+#include <tuple>
+
+namespace Fortran::semantics {
+
+template <typename T> class CaseValues {
+public:
+  CaseValues(SemanticsContext &c, const evaluate::DynamicType &t)
+      : context_{c}, caseExprType_{t} {}
+
+  void Check(const std::list<parser::CaseConstruct::Case> &cases) {
+    for (const parser::CaseConstruct::Case &c : cases) {
+      AddCase(c);
+    }
+    if (hasErrors_) {
+      return;
+    }
+    cases_.sort(Comparator{});
+    auto endIter{cases_.end()};
+    bool disjoint{true};
+    for (auto iter{cases_.begin()}; iter != endIter; ++iter) {
+      if constexpr (T::category == TypeCategory::Logical) { // C1148
+        if (!iter->IsDefault() &&
+            (!iter->lower || !iter->upper || *iter->lower != *iter->upper)) {
+          context_.Say(iter->stmt.source,
+              "CASE range is not allowed for LOGICAL"_err_en_US);
+        }
+      }
+      if (iter->lower && iter->upper && *iter->lower > *iter->upper) {
+        context_.Say(iter->stmt.source,
+            "CASE %s has lower bound greater than upper bound"_en_US,
+            iter->AsFortran());
+      }
+      // C1149: Ensure each case in the sorted list is disjoint from the next.
+      auto next{iter};
+      if (++next != endIter && !Comparator{}(*iter, *next)) {
+        disjoint = false;
+      }
+    }
+    if (!disjoint) {
+      ReportConflictingCases();
+    }
+  }
+
+private:
+  using Value = evaluate::Scalar<T>;
+
+  void AddCase(const parser::CaseConstruct::Case &c) {
+    const auto &stmt{std::get<parser::Statement<parser::CaseStmt>>(c.t)};
+    const parser::CaseStmt &caseStmt{stmt.statement};
+    const auto &selector{std::get<parser::CaseSelector>(caseStmt.t)};
+    std::visit(
+        common::visitors{
+            [&](const std::list<parser::CaseValueRange> &ranges) {
+              for (const auto &range : ranges) {
+                cases_.emplace_back(stmt);
+                cases_.back().ComputeBounds(*this, range);
+              }
+            },
+            [&](const parser::Default &) { cases_.emplace_front(stmt); },
+        },
+        selector.u);
+  }
+
+  std::optional<Value> GetValue(const parser::CaseValue &caseValue) {
+    const auto &expr{caseValue.thing.thing.value()};
+    auto *x{expr.typedExpr.get()};
+    if (x && x->v) { // C1147
```

Comment:
Using `if (auto *x{GetExpr(caseValue)}) {` here might simplify things, but I'm not sure about that.

Reply:
Will try it out.